### PR TITLE
ROX-22019: Protobuf v2 migration

### PR DESCRIPTION
**Pull request by mtodor**
This PR migrates scanner from gogo protobuf V1 to protobuf V2 stack.

To simplify the review, diffs can checked per commit.

Commit `Change proto gen tools to protobuf V2`:
- changes proto generation tools to V2 tools
- split generated `.pb.go` file into multiple files:
- - `.pb.go` - contains structs
- - `*_vtproto.pb.go` - contains vtproto generated code. Primarily marshaling/unmarshalling, etc.
- - `*_grpc.pb.go` - contains gRCP bindings code
- added 3rd party proto files. They have been part of `github.com/grpc-ecosystem/grpc-gateway` before, but they are not included anymore.

Commit `Generate code from proto files`:
- contains only generated code

Commit `Adjust code for protobuf V2 generated code`:
- change to gRPC Gateway V2
- replace old gogo proto types with the protobuf V2 ones

### CODE REVIEW

File: make/protogen.mk [hunk 1]
```diff
@@ -14,13 +14,14 @@ ALL_PROTOS_REL = $(ALL_PROTOS:$(PROTO_BASE_PATH)/%=%)
...
 
 GENERATED_BASE_PATH = $(BASE_PATH)/generated
 GENERATED_DOC_PATH = image/docs
 MERGED_API_SWAGGER_SPEC = $(GENERATED_DOC_PATH)/api/v1/swagger.json
 GENERATED_API_DOCS = $(GENERATED_DOC_PATH)/api/v1/reference
 GENERATED_PB_SRCS = $(ALL_PROTOS_REL:%.proto=$(GENERATED_BASE_PATH)/%.pb.go)
+GENERATED_VT_SRCS = $(ALL_PROTOS_REL:%.proto=$(GENERATED_BASE_PATH)/%_vtproto.pb.go)
+GENERATED_API_SRCS = $(SERVICE_PROTOS_REL:%.proto=$(GENERATED_BASE_PATH)/%_grpc.pb.go)
 GENERATED_API_GW_SRCS = $(SERVICE_PROTOS_REL:%.proto=$(GENERATED_BASE_PATH)/%.pb.gw.go)
 GENERATED_API_SWAGGER_SPECS = $(API_SERVICE_PROTOS:%.proto=$(GENERATED_BASE_PATH)/%.swagger.json)
```
Comment (RTann) [inline make/protogen.mk:26]: if you're interested, we can probably remove anything related to this, too. I don't think we generate swagger specs

File: pkg/clairify/server/server.go [hunk 2]
```diff
@@ -257,14 +257,8 @@ func (s *Server) GetVulnDefsMetadata(w http.ResponseWriter, _ *http.Request) {
...
 
-	ts, err := protoTypes.TimestampProto(t)
-	if err != nil {
-		clairErrorString(w, http.StatusInternalServerError, "failed to obtain vulnerability definitions update timestamp: %v", err)
-		return
-	}
-
 	vulnDefsInfo := &protoV1.VulnDefsMetadata{
-		LastUpdatedTime: ts,
+		LastUpdatedTime: timestamppb.New(t),
```
[the remark this answers is not in the document]
Comment (mtodor) [reply pkg/clairify/server/server.go:261]: Nice find! I missed the fact that validation was done within `TimestampProto` before.

File: api/grpc/grpc.go [hunk 3]
```diff
@@ -190,7 +190,7 @@ func (a *apiImpl) muxer(localConn *grpc.ClientConn) http.Handler {
 		mux.Handle(route, handler)
 	}
 
-	gwMux := runtime.NewServeMux(runtime.WithMarshalerOption(runtime.MIMEWildcard, &runtime.JSONPb{EmitDefaults: true}))
```
[the remark this answers is not in the document]
Comment (mtodor) [reply api/grpc/grpc.go:193]: We are using `github.com/grpc-ecosystem/grpc-gateway/v2/runtime` (`v2`) now.

And they have changed structs: https://grpc-ecosystem.github.io/grpc-gateway/docs/development/grpc-gateway_v2_migration_guide/#we-now-emit-default-values-for-all-fields

File: make/protogen.mk (hunk 1 above)
Comment (mtodor) [reply make/protogen.mk:26]: I would keep that out. There are too many changes for my taste to couple it with protobuf V2 migration.

That cleanup can be done before (I don't mind rebasing) or after protobuf V2 migration, but as a separate PR.

File: third_party/googleapis/google/api/http.proto
Comment (RTann) [inline third_party/googleapis/google/api/http.proto:1]: I forgot to ask before, but why do we need this? Does this not exist in proto v2?

File: e2etests/vuln_test.go [hunk 4]
```diff
@@ -6,10 +6,10 @@ package e2etests
 import (
 	"context"
 	"fmt"
+	"slices"
 	"strings"
 	"testing"
 
-	"github.com/stackrox/rox/pkg/sliceutils"
```
Comment (RTann) [inline e2etests/vuln_test.go:12]: big fan of removing this

File: make/protogen.mk
Comment (RTann) [inline make/protogen.mk:1]: I'm definitely not an expert for this, but based on the fact that the generated files were created, as long as CI passes, then LGTM

**Review by RTann — APPROVED** [review]
Test failure looks unrelated. LGTM thanks

File: third_party/googleapis/google/api/http.proto
Comment (mtodor) [reply third_party/googleapis/google/api/http.proto:1]: Yes. This is removed from `github.com/grpc-ecosystem/grpc-gateway/v2`. And `v1` had it. https://github.com/grpc-ecosystem/grpc-gateway/tree/33ca56eeaf85003043d90915d01d79a95dae870c?tab=readme-ov-file#2-with-custom-annotations

There are other libraries that still include these files, but that is not reliable. I think it's best to copy that in our repo.

File: go.mod [hunk 5]
```diff
@@ -182,16 +185,10 @@ require (
...
-	// - generate MessageClone() and Clone() functions
-	// - error on nil unmarshal instead of panic
-	// - trailing comments propagation to generated code
-	github.com/gogo/protobuf => github.com/stackrox/protobuf v1.2.2-0.20240207122816-e936d453291c
-
 	github.com/heroku/docker-registry-client => github.com/stackrox/docker-registry-client v0.1.0
 
 	github.com/operator-framework/helm-operator-plugins => github.com/stackrox/helm-operator v0.0.12-0.20230825152000-1361e2f7db46
-	github.com/stackrox/rox => github.com/stackrox/stackrox v0.0.0-20240402171531-15fa6d254174
+	github.com/stackrox/rox => github.com/stackrox/stackrox v0.0.0-20240510180132-d8b1b7d8d248
```
Comment (janisz) [inline go.mod:191]: I think it would be good to update it to latest master version in a separated PR to minimize diff.
Also update to master could be merged right now.

**Review by janisz — APPROVED** [review]

File: go.mod [hunk 6]
```diff
@@ -1,6 +1,6 @@
 module github.com/stackrox/scanner
 
-go 1.21
+go 1.21.8
```
Comment (RTann) [inline go.mod:3]: Is this change required?

File: go.mod [hunk 7]
```diff
@@ -202,7 +199,7 @@ replace (
 	github.com/mholt/archiver/v3 => github.com/anchore/archiver/v3 v3.5.2
 
 	github.com/operator-framework/helm-operator-plugins => github.com/stackrox/helm-operator v0.0.12-0.20230825152000-1361e2f7db46
-	github.com/stackrox/rox => github.com/stackrox/stackrox v0.0.0-20240402171531-15fa6d254174
+	github.com/stackrox/rox => github.com/stackrox/stackrox v0.0.0-20240711164643-43f1b56f2130
```
Comment (RTann) [inline go.mod:202]: Is the plan to change this to the master branch before or after merging this PR?

File: e2etests/vuln_test.go [hunk 8]
```diff
@@ -6,10 +6,10 @@ package e2etests
 import (
 	"context"
 	"fmt"
+	"slices"
```
Comment (janisz) [inline e2etests/vuln_test.go:9]: This change is not related to migration

File: make/protogen.mk [hunk 9]
```diff
@@ -168,61 +159,93 @@ printprotos:
...
 	@echo "+ $@"
 	$(SILENT)mkdir -p $(dir $@)
 	$(SILENT)PATH=$(PROTO_GOBIN) $(PROTOC) \
-		-I$(GOGO_DIR) \
 		-I$(PROTOC_INCLUDES) \
-		-I$(GRPC_GATEWAY_DIR)/third_party/googleapis \
+		-I$(GOOGLE_API_INCLUDES) \
 		--proto_path=$(PROTO_BASE_PATH) \
-		--gofast_out=$(GOGO_M_STR:%=%,)$(M_ARGS_STR:%=%,)plugins=grpc:$(GENERATED_BASE_PATH) \
+		--plugin protoc-gen-go-grpc="${PROTOC_GEN_GO_GRPC}" \
```
Comment (janisz) [inline make/protogen.mk:222]: Since we pass full path to plugin binary do we need to  change `$PATH` in 218?

**Review by janisz — APPROVED** [review]

File: go.mod (hunk 7 above)
Comment (mtodor) [reply go.mod:202]: @RTann After merging. My plan is to merge both PRs (at the same time) and then bump the versions in both repos to use the `master` branch versions.

File: go.mod (hunk 6 above)
Comment (mtodor) [reply go.mod:3]: This is updated by `go mod tidy`. I'm not sure about the reasons, it could be because of `stackrox/stackrox` - https://github.com/stackrox/stackrox/blob/195784f810aea44b077f9bda82851c91eb84b45f/go.mod#L3

File: e2etests/vuln_test.go (hunk 8 above)
Comment (mtodor) [reply e2etests/vuln_test.go:9]: @janisz Here is PR to pull that out: https://github.com/stackrox/scanner/pull/1572

File: make/protogen.mk (hunk 9 above)
Comment (mtodor) [reply make/protogen.mk:222]: We can do that. I would do that as a follow-up and do a cleanup of `make/protogen.mk`. All swagger.json targets can be removed.

File: go.mod (hunk 6 above)
Comment (RTann) [reply go.mod:3]: makes sense

File: e2etests/vuln_test.go (hunk 8 above)
Comment (mtodor) [reply e2etests/vuln_test.go]: Changes for `slices` are removed.

**Review by RTann — APPROVED** [review]
